### ISSUE TITLE
Migrate to Sass module system (@use)

### ISSUE THREAD

**Issue by 36degrees** [alphagov/govuk-frontend#1791]
## What

Update GOV.UK Frontend to use the new `@use` syntax internally to import dependencies, and to be consumed using `@use`.

## Why

The Sass [module system](https://sass-lang.com/blog/the-module-system-is-launched) has been introduced, aiming to address some of the issues with the `@import` rule.

The `@import` rule, and most global functions will eventually be deprecated, and then removed from Sass.

Moving to `@use` will allow us to revisit and simplify the way that GOV.UK Frontend is implemented, without the limitations of `@import – for example namespacing, and the [govuk-exports](https://github.com/alphagov/govuk-frontend/blob/42742381f5f231b6d6a9962f6baddfcf15710efb/src/govuk/tools/_exports.scss) mixin.

## Depends on

Stable support for [`@uses` in LibSass](https://github.com/sass/libsass/issues/2807) – which probably means a stable version with support having been out for 3-6 months.

(The ['future plans' section of the module system blog post](https://sass-lang.com/blog/the-module-system-is-launched#future-plans) sets a hard limit of October 2021 for deprecating `@import`, which we could also do, but we'd need to have a better understanding of the impact it'll have on our users.)

## Further detail

Our [supported Sass compilers](https://frontend.design-system.service.gov.uk/installing-with-npm/#requirements) would then be:

- Dart Sass 1.23.0 or higher
- LibSass 5.0 or higher (based on the tracked release in https://github.com/sass/libsass/issues/2807)

We would drop support for Ruby Sass, which is deprecated (as of 26 March 2019) and will not be updated to use the new module system.

**Comment by 36degrees** [comment]
It looks as though LibSass is now being deprecated:
https://github.com/sass/libsass/issues/3123

**Comment by 36degrees** [comment]
The beta of 2.10.0 of the US Web Design System migrates to the module system:

https://designsystem.digital.gov/about/releases/#version-2100-beta-0

**Comment by thisisdano** [comment]
From the USWDS side, we're finding that the modules support is very patchy — perhaps too patchy to support in the near term. When we can control the compile chain, we can feel confident using modules, but we have a lot of project users that compile using, say, the built-in compiler in Jekyll, and this cannot parse modules at all. What I believe we've found — though this is not yet final — is that while we intend to move to modules when we can, we may strip this functionality from the beta and go as far as we can using the old syntax, waiting to fully move to modules if and when we find better overall support.

**Comment by 36degrees** [comment]
Ah that's really helpful to know @thisisdano – thanks so much for taking the time to pass on your findings 🙏🏻  I think we are still a fair way of actually being able to migrate for similar reasons.

**Comment by andymantell** [comment]
In case it's useful to you in future - we found on nhsuk-frontend that switching to dart-sass resulted in surprisingly significant rendering differences due to the increased default precision in dart-sass. That precision isn't configurable with dart-sass, and we can't control everyone's build chains anyway, so we have gone with the following solution for the moment:

https://github.com/nhsuk/nhsuk-frontend/pull/731

Basically we are forcibly rounding the line height to 5 decimal places so that rendering is consistent between node-sass and dart-sass.

It's worth noting that anyone currently consuming govuk-frontend and using dart-sass will already be getting differences in rendering. If that concerns you, it might be worth putting a similar fix in ahead of any migration to dart-sass.

**Comment by querkmachine** [comment]
I've been digging around the Sass repos for new information on when `@import` is going to be deprecated or removed, given there hasn't been any official update that the date was changed from October 2021. 

The comment here (December 2021) seems to imply that the deprecation did not take place due to the ecosystem not migrating to Dart Sass as quickly as anticipated, with no new date being set for it to take place. https://github.com/sass/sass/issues/3223#issuecomment-1002813284

This follow up I've asked indicates that they consider 70–80% adoption of Dart Sass to be the threshold for them to start the deprecation process. https://github.com/sass/sass/issues/3276#issuecomment-1076789429

**Comment by querkmachine** [comment]
Apparently Dart Sass adoption is [currently around 66%](https://github.com/sass/sass/issues/3276#issuecomment-1090803953).

**Comment by 36degrees** [comment]
[Sass' 'The Module System is Launched' blog post](https://sass-lang.com/blog/the-module-system-is-launched) was amended in July 2022 with this update:

> In light of the fact that LibSass was deprecated before ever adding support for the new module system, the timeline for deprecating and removing `@import` has been pushed back. We now intend to wait until 80% of users are using Dart Sass (measured by npm downloads) before deprecating `@import`, and wait at least a year after that and likely more before removing it entirely.

Looking at download stats we're getting very close to that 80% mark (79.5% or 79.64% if you also include downloads of sass-embedded), so we could see `@import` deprecated soon. However, we'll still have at least a year before it's removed entirely.

**Comment by querkmachine** [comment]
Dart Sass adoption is, as of today, at around 83.96%. 

It looks like the Sass team have started preparing to deprecate `@import`, but on a slower trajectory than was originally described on the original blog post. https://github.com/sass/sass/issues/3513

**Comment by querkmachine** [comment]
It has now been specified that `@import` will be deprecated before the end of 2024 in conjunction with the release of Dart Sass 2.0.0. You can see progress on that in the milestone. https://github.com/sass/dart-sass/milestone/3

**Comment by romaricpascal** [comment]
> It's now been stated that the official deprecation is likely to happen in the next few weeks.

'Next few weeks' have turned into months, but [the deprecation has finally happened in Sass 1.80](https://sass-lang.com/blog/import-is-deprecated/).

**Comment by sussexrick** [comment]
We consume the GOV.UK SASS files along the following lines:

```scss
@import '_our-variables.scss';  // sets new values for some GOV.UK variables
@import 'govuk';                // imports GOV.UK Frontend, using new variable values
@import '_our-styles.scss';     // adds other customisation, eg @extend and override your selectors
```

I've just tried migrating our code to `@use` and had to roll back all the changes because it didn't compile as expected. Our overrides didn't work and we got base GOV.UK styling. I'm just commenting here to flag it as a use case for when you do come to make this change.

**Comment by querkmachine** [comment]
@sussexrick This is likely because the Sass module system no longer automatically exports variables, functions and mixins to Sass's global scope. This means that `govuk` is unable to access anything in `_our-variables.scss` and `_our-styles.scss` is unable to access anything from `govuk`.

In the interim, you can force things to be exported to the global scope using `@use … as`:

```scss
@use '_our-variables.scss' as *;
@use 'govuk' as *;
@use '_our-styles.scss' as *;
```

This is likely to change In future when we begin implementing the module system. This is likely (though not guaranteed) to make use of `@use … with`.

```scss
@use 'govuk' with (
  // Customisations move to here, for example...
  $govuk-brand-colour: rebeccapurple,
  $govuk-page-width: 1080px
);
@use '_our-styles.scss';
```

The `_our-styles.scss` file would need to be updated to explicitly import the specific parts of govuk-frontend that it uses, rather than relying on them existing in the global scope. For example: 

```scss
@use 'govuk/helpers/colour'; // automatically gets namespaced as 'colour'
@use 'govuk/helpers/typography'; // automatically gets namespaced as 'typography'

.custom-thing {
  color: colour.govuk-colour('white'); 
  background-color: colour.govuk-colour('blue'); 
  @include typography.govuk-font-size(19);
}
```

These latter examples are still up in the air, as we haven't actually implemented the module system yet, so it's subject to change and probably won't work quite as expected at the moment. Hopefully it's useful to see the direction things might move in!

**Comment by sussexrick** [comment]
Thank you for the suggestion but the interim code doesn't work. If `_our-variables.scss` contains this:

```scss
$govuk-font-family: 'Open Sans', Arial, sans-serif;
```

...then using `@import` all the fonts in GOV.UK classes are updated. Using `@use` as shown above they are not.

I'll wait for your official support and follow the guidance that comes out at that time. Please bear this use case in mind though. You've done a great job of making things customisable up to now. 👍

**Comment by romaricpascal** [comment]
As of [GOV.UK Frontend 6.2.0](https://github.com/alphagov/govuk-frontend/releases/v6.2.0), we support Sass modules 🥳
